x/programs/examples/imports/program/program.go

```diff
@@ -170,44 +159,54 @@ func (i *Import) callProgramFn(
 	return int64(res[0])
 }
 
-func getCallArgs(ctx context.Context, memory runtime.Memory, buffer []byte, invokeProgramID int64) ([]int64, error) {
+// getCallArgs returns the arguments to be passed to the program being invoked from [buffer].
+func getCallArgs(ctx context.Context, memory runtime.Memory, buffer []byte, programAddressBytes []byte) ([]runtime.SmartPtr, error) {
 	// first arg contains id of program to call
-	args := []int64{invokeProgramID}
-	p := codec.NewReader(buffer, len(buffer))
-	i := 0
-	for !p.Empty() {
-		size := p.UnpackInt64(true)
-		isInt := p.UnpackBool()
-		if isInt {
-			valueInt := p.UnpackInt64(true)
-			args = append(args, valueInt)
-		} else {
-			valueBytes := make([]byte, size)
-			p.UnpackFixedBytes(int(size), &valueBytes)
-			ptr, err := runtime.WriteBytes(memory, valueBytes)
-			if err != nil {
-				return nil, err
-			}
-			args = append(args, ptr)
-		}
-		i++
+	invokeProgramAddressPtr, err := runtime.WriteBytes(memory, programAddressBytes)
+	if err != nil {
+		return nil, err
 	}
-	if p.Err() != nil {
-		return nil, fmt.Errorf("failed to unpack arguments: %w", p.Err())
+	argPtr, err := runtime.NewSmartPtr(uint32(invokeProgramAddressPtr), len(programAddressBytes))
+	if err != nil {
+		return nil, err
 	}
+
+	args := []runtime.SmartPtr{argPtr}
+
+	for i := 0; i < len(buffer); {
+		// unpacks uint32
+		lenBytes := buffer[i : i+consts.Uint32Len]
+		length := binary.BigEndian.Uint32(lenBytes)
+
+		valueBytes := buffer[i+consts.Uint32Len : i+consts.Uint32Len+int(length)]
+		i += int(length) + consts.Uint32Len
+
+		// every argument is a pointer
+		ptr, err := runtime.WriteBytes(memory, valueBytes)
+		if err != nil {
+			return nil, err
+		}
+		argPtr, err := runtime.NewSmartPtr(uint32(ptr), int(length))
+		if err != nil {
+			return nil, err
+		}
+		args = append(args, argPtr)
+	}
+
 	return args, nil
 }
 
-func getProgramWasmBytes(log logging.Logger, db state.Immutable, idBytes []byte) ([]byte, error) {
-	id, err := ids.ToID(idBytes)
-	if err != nil {
-		return nil, err
+func getProgramWasmBytes(log logging.Logger, db state.Immutable, addressBytes []byte) ([]byte, error) {
+	if len(addressBytes) != codec.AddressLen {
+		return nil, fmt.Errorf("invalid address length: %d", len(addressBytes))
 	}
 
+	address := codec.Address(addressBytes)
```

Comment:
You are probably going to want to assert that the first byte is the right typeID (in the VM setting).

Reply:
to clarify, this should be done in the VM? Not sure how the engine(at least in our current implementation) can ensure the right typeID's here. 
More generally, the SDK needs to handle address and `Program` types in a more tamper-proof way, but I'm still looking into that.

Reply:
i guess, is this comment suggesting I should address(no pun intended) something now or more of a note for when we start building out the VMs later?

Reply:
> to clarify, this should be done in the VM? Not sure how the engine(at least in our current implementation) can ensure the right typeID's here. More generally, the SDK needs to handle address and `Program` types in a more tamper-proof way, but I'm still looking into that.

You are right. This should be handled by the wrapping VM. Can you leave a comment/invariant stating that?